utils/s3/client.cc

```diff
@@ -854,6 +861,209 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {}, {} bytes (upload id {})", part_number, part_size, _upload_id);
+        req.write_body("bin", part_size, [f=std::move(f), offset, part_size, this] (output_stream<char>&& out_) mutable -> future<> {
+            auto output = std::move(out_);
+            auto input = make_file_input_stream(std::move(f), offset, part_size);
+            co_await input.consume(throttled_consumer{output, *_client});
+            co_await output.close();
+            co_await input.close();
+        });
+        return _client->make_request(std::move(req), [this, part_size, part_number, start = s3_clock::now()] (group_client& gc, const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+            auto etag = reply.get_header("ETag");
+            s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+            _part_etags[part_number] = std::move(etag);
+            gc.write_stats.update(part_size, s3_clock::now() - start);
+            return make_ready_future();
+        }).handle_exception([this, part_number] (auto ex) {
+            s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+        });
+    }
+
+    static size_t calc_part_size(size_t total_size, size_t part_size) {
+        if (part_size > 0) {
+            assert(part_size >= aws_minimum_part_size);
+            assert((total_size + part_size - 1) / part_size <= aws_maximum_parts_in_piece);
+            return part_size;
+        }
+        // if part_size is 0, this means the caller leaves it to us to decide
+        // the part_size. and let's start with aws_minimum_part_size. and we
+        // prefer smaller parts, so we increment it by 1M  until the number of
+        // parts satisfies the constraint of aws_maximum_parts_in_piece
+        part_size = aws_minimum_part_size;
+        while ((total_size + part_size - 1) / part_size > aws_maximum_parts_in_piece) {
+            part_size += 1_MiB;
+        }
+        return part_size;
+    }
+
+    future<> multi_part_upload(file&& f, uint64_t total_size, size_t part_size) {
+        co_await create_multipart_upload();
+
+        for (size_t offset = 0; offset < total_size; offset += part_size) {
+            part_size = std::min(total_size - offset, part_size);
+            s3l.trace("upload_part: {}~{}/{}", offset, part_size, total_size);
+            // upload the parts in the background for better throughput
+            auto gh = _bg_uploads.hold();
```

Comment:
This place is a bit worrisome. It might create up to 10000 fibers calling upload_part(). Eventually most of those will stuff-up the client's connections pool limit thus blocking this sched group's client from _any_ activity until the upload finishes (the client's queue uses cond-var waiting which is FIFO) .

Opposite to that, regular sink and jumbo sink both first claim S3 memory, then go and call make_request()

Reply:
indeed. i thought about this, and had the same concern. how about adding yet another semaphore? and its size is configured by the caller, or by `client` ?

Reply:
another way around is to have a loop like (in pseudo code)
```py
while (remaining < total_size) {
  auto chunk_size = client.options.chunk_size
  auto units = co_await client.claim_memoy(chunk_size);
  auto chunk = co_await input.read(chunk_size); 
  auto chunk_size = chunk.size();
  co_await output.write(std::move(chunk));
  remaining -= chunk_size;
}
```

but the downside is that we cannot use `copy()` anymore.

Reply:
> but the downside is that we cannot use copy() anymore.

Which copy()? I don't see any in this PR

Reply:
sorry, i meant `input.consume(throttled_consumer{output, *_client});` . it mirrors the implementation of `copy()` in seastar.

Reply:
Ah. Yes :( that's unfortunate